packages/stream-node/writer.js

```diff
@@ -0,0 +1,36 @@
+/* Adapts a Node.js stream to an Writer<Uint8Array>, where a writer stream is
+ * modeled as a hybrid async iterator + generator.
+ */
+
+// @ts-check
+/// <reference types="ses"/>
+
+/**
+ * Adapts a Node.js writable stream to a JavaScript
+ * async iterator of Uint8Array data chunks.
+ * Back pressure emerges from awaiting on the promise
+ * returned by `next` before calling `next` again.
+ *
+ * @param {NodeJS.WritableStream} writer the destination Node.js writer
+ * @returns {import('@endo/stream').Writer<Uint8Array, void>}
+ */
+export const makeNodeWriter = writer => {
+  return harden({
+    async next(value) {
+      return new Promise(resolve => {
+        if (!writer.write(value)) {
+          writer.once('drain', resolve);
+        } else {
+          resolve();
```

Comment:
I don't understand the `else` case.

Reply:
Node is a little peculiar when it comes to writable streams. The model is that you call `write`, and if it returns `false`, you should no longer call write until the `'drain'` event has been dispatched. By resolving right away it informs the writer that further writing calls can happen, preserving the node stream semantics.

Reply:
Also, as Mathieu answers, writer.write tells the caller whether the underlying buffer is above a high water mark. It’s not a ring buffer, so it doesn’t have to return a partial success, but it does advise you to back off until drain.

I’m translating that to a system where `next` returns a promise that resolves when you can safely send more data. If `write` returns true, that’s immediate.

An alternate framing would return a promise for the drain even only if write returns false, but I vaguely suspect that always returning a promise is better.

```js
if (!writer.write(value)) {
  return new Promise(resolve => writer.once('drain', resolve));
}
```

Reply:
That wouldn't help since `next` has to be an async function that always returns a Promise.